src/libsync/propagatebundle.cpp

```diff
@@ -0,0 +1,602 @@
+        _jobs.append(job);
+        connect(job, SIGNAL(finishedSignal()), this, SLOT(slotMultipartFinished()));
+        connect(job, SIGNAL(destroyed(QObject*)), this, SLOT(slotJobDestroyed(QObject*)));
+        job->start();
+        _preparingBundle = false;
+
+        // check if there are any items to be bundled and if you can run another parallel job
+        if (!_itemsToChecksum.empty() && (_propagator->_activeJobList.count() < _propagator->maximumActiveJob())) {
+            start();
+        }
+    } else {
+        delete job;
+    }
+}
+
+QByteArray PropagateBundle::getRemotePath(QString filePath){
+    QString remotePath(_propagator->_remoteFolder+filePath);
+    return remotePath.toStdString().c_str();
+}
+
+void PropagateBundle::append(const SyncFileItemPtr &bundledFile){
+    _size += bundledFile->_size;
+    _itemsToChecksum.append(bundledFile);
+}
+
+bool PropagateBundle::empty(){
+    return _itemsToChecksum.empty();
+}
+
+void PropagateBundle::slotMultipartFinished()
+{
+    MultipartJob *job = qobject_cast<MultipartJob *>(sender());
+    Q_ASSERT(job);
+    slotJobDestroyed(job); // remove it from the _jobs list
+
+    qDebug() << Q_FUNC_INFO << job->reply()->request().url() << "FINISHED WITH STATUS"
+             << job->reply()->error()
+             << (job->reply()->error() == QNetworkReply::NoError ? QLatin1String("") : job->reply()->errorString())
+             << job->reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute)
+             << job->reply()->attribute(QNetworkRequest::HttpReasonPhraseAttribute);
+
+    _propagator->_activeJobList.removeOne(this);
+
+    QNetworkReply::NetworkError err = job->reply()->error();
+
+#if QT_VERSION < QT_VERSION_CHECK(5, 4, 2)
+    if (job->reply()->error() == QNetworkReply::OperationCanceledError && job->reply()->property(owncloudShouldSoftCancelPropertyName).isValid()) {
+        // Abort the job and try again later.
+        // This works around a bug in QNAM wich might reuse a non-empty buffer for the next request.
+        qDebug() << "Forcing job abort on HTTP connection reset with Qt < 5.4.2.";
+        _propagator->_anotherSyncNeeded = true;
+        abortWithError(SyncFileItem::SoftError, tr("Forcing job abort on HTTP connection reset with Qt < 5.4.2."));
+        return;
+    }
+#endif
+
+    if (err != QNetworkReply::NoError) {
+        _item->_httpErrorCode = job->reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+
+        QByteArray replyContent = job->reply()->readAll();
+        qDebug() << replyContent; // display the XML error in the debug
+        QString errorString = errorMessage(job->errorString(), replyContent);
+
+        if (job->reply()->hasRawHeader("OC-ErrorString")) {
+            errorString = job->reply()->rawHeader("OC-ErrorString");
+        }
+
+        //TODO: soft error till other mechanism will be added - this will retry bundle later.
+        abortWithError(SyncFileItem::SoftError, errorString);
```

Comment:
@ogoffart @jturcotte For now I am just using SoftError in case of any full Bundle Errors. How can I implement other mechanism for whole the bundle? I mean, the problem here that currently we blacklist by filename, not by other means. How can I blacklist bundle ? I mean, this will be good mechanism on weak networks, if this will fail for some reason (airport network stops working), the client will roll back immedietaly to standard PUTs, not to try upload like 10+ files again.